A modular rewrite featuring numerous enhancements and bug fixes

9ao9ai9ar opened the pull request:
The only downsides of my rewrite that I can think of are the higher memory usage (by about 50%) and increased dependencies.

9ao9ai9ar committed:
A modular rewrite with tangible improvements for both devs and users.

Some common development practices are introduced to the project:
* data (de)serialization and validation, via Pydantic
* API client code generation, via OpenAPI
* continuous integration, via a suite of tests and utilities

Known bugs like posts on meta sites weren't being backed up and #1 are
fixed. More consideration is given to rate limiting while driving down
quota usage by using static filters; a few CLI options are also added.

mhdadk commented:
Hey @9ao9ai9ar, thank you very much for this PR! It seems like you really cleaned up the code. Adding tests was an especially nice touch.

Before merging this PR, because it is quite large (49 files changed), I'd like to review it in more detail to make sure it is OK. Unfortunately, I'm not sure when I will have the time in the near future to do this. I'm going to keep this PR open until I do so.

Nevertheless, I wanted to write to you to give you some positive feedback on your work.

9ao9ai9ar commented:
Thanks for your words of encouragement. I will be pushing out minor commits in the meantime, like updating the README and upgrading dependencies (Pydantic v2.9.0 is just around the corner, promising to fix the high memory consumption).

I do have an idea to improve the output format, but I will withhold work on that until next year. The idea is to first move all non-body fields of the posts into the (non-standard but widely supported) YAML metadata section at the beginning of the Markdown files. Secondly, individual answers should probably be separated into individual files to keep them as close to the original as possible without modifying the Markdown content by inserting the section markers. Lastly, the licenses of the posts (currently there are 3 versions of CC in use) should be included and linked. I can open an issue if you're interested in further discussion.

mhdadk commented:
All these changes sounds good to me. I can have a look at them as well when you push them (and when I have some more time).

9ao9ai9ar commented:
I have three news to share with you, one good, one bad, and one neutral:

1. I've thought of another important feature to add to this project: include an option to download all images as files or base64 encoded data blobs. In the data dumps, only the image URLs are saved, so there is still a risk of URL rot and losing the pictures. I can probably start the work in the second half of the year.
2. The "slight" increase in dependencies has negatively affected the ability to run the script on some platforms. For example, I wasn't able to get the project to build on NetBSD due to compile errors in PyO3. Pydantic 2.9 also didn't fix the high memory consumption; they are tackling the memory issues again in Pydantic 2.11, but if no progress were made either in this release, I might have to reconsider the choice of incorporating this library in the code, which would be a shame as I kind of like their API.
3. Since this is more a rewrite than a refactor, it's less important to do piecemeal commits in my humble opinion, so as I intend to do a git rebase to squash the commits into one, I will mark this pull request as draft for the time being. In the future, I will open feature branches to work on and do the rebases there.

mhdadk commented:
> I have three news to share with you, one good, one bad, and one neutral:
> 
>     1. I've thought of another important feature to add to this project: include an option to download all images as files or base64 encoded data blobs. In the data dumps, only the image URLs are saved, so there is still a risk of URL rot and losing the pictures. I can probably start the work in the second half of the year.
> 
>     2. The "slight" increase in dependencies has negatively affected the ability to run the script on some platforms. For example, I wasn't able to get the project to build on NetBSD due to compile errors in PyO3. Pydantic 2.9 also didn't fix the high memory consumption; they are tackling the memory issues again in Pydantic 2.11, but if no progress were made either in this release, I might have to reconsider the choice of incorporating this library in the code, which would be a shame as I kind of like their API.
> 
>     3. Since this is more a rewrite than a refactor, it's less important to do piecemeal commits in my humble opinion, so as I intend to do a git rebase to squash the commits into one, I will mark this pull request as draft for the time being. In the future, I will open feature branches to work on and do the rebases there.

Nice work!

9ao9ai9ar committed:
Fix type mismatch in generated models due to typos.

9ao9ai9ar committed:
New output directory and file layouts.

All data except the post bodies are placed in the YAML front matter.
The metadata now include the necessary attributions and links as
required by the CC BY-SA licenses. Auto translation of line separator
characters has been turned off, so no more extra newlines breaking
Markdown syntax.

9ao9ai9ar committed:
Refactor codebase.

9ao9ai9ar committed:
Add paging to read-filter to unify behaviors and unbreak things.

This may be a documentation bug because read-filter actually supports
paging, and no other vectorized APIs omit the paging parameters.

9ao9ai9ar committed:
Switch to reputable, pure Python libraries.

To offer stronger guarantees that the code will run in other
implementations of Python, and to avoid pesky compilation steps on
lesser platforms, we switch to select pure Python libraries after a
round of evaluation of the available options. The results of this
switch, compared with both the prior Pydantic implementation and the
original dictionary-based implementation by mhdadk, and in terms of
the total dependency count and peak memory usage when run on a dummy
account (ACCOUNT_ID = 8), are as follows:
* dict: 5 dependencies; 41 MB
* Pydantic: 21 dependencies; 60 MB
* attrs: 10 dependencies; 48 MB

All typing issues are fixed: the suppressed warnings are either false
positives or ignorable. Detection of self-answered questions has been
restored to past performance before the output format redesign. There
is cleaner separation of concerns and more additions to CLI options in
anticipation of certain use cases.

9ao9ai9ar committed:
Add missing required dependency.

9ao9ai9ar committed:
Other types of errors may be raised in case of timestamp overflow.

9ao9ai9ar committed:
Rename module.

9ao9ai9ar committed:
Update support status and apply the latest pylint's suggestions

9ao9ai9ar committed:
Adjust tool configs.

9ao9ai9ar committed:
Update to the new authentication scheme to prevent future blockage.

9ao9ai9ar committed:
Maintenance release and project roadmap.
